_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,227 @@
+
+Cruise Control ships its own custom metric reporter, the `Cruise Control metric reporter` but must be included in `/opt/kafka/libs/` directory of every Kafka broker in order for it to work. For making the Cruise Control metric reporter available to Strimzi Kafka brokers, one can create a new Kafka image from the following Dockerfile:
+
+```Dockerfile
+FROM strimzi/kafka:0.11.4-kafka-2.0.1
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"
+
+USER root
+
+RUN yum -y install git java-1.8.0-openjdk-devel \
+    && yum clean all -y
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+RUN cd cruise-control \
+    && ./gradlew jar :cruise-control-metric-reporter:jar \
+    && cp cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
+```
+
+Build the image:
+
+```bash
+docker build . -t <kafka-image-name>
+```
+
+Then create a basic Kafka resource referencing the new Kafka image and specifying the `CruiseControlMetricsReporter` in the Kafka config one of the `metrics.reporters` using the following `kafka.yaml` file:
+```yaml
+apiVersion: kafka.strimzi.io/v0alpha1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    image: <kafka-image-name>
+    version: 2.0.1
+    replicas: 3
+    listeners:
+      plain: {}
+      tls: {}
+    config: 
+      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
+      offsets.topic.replication.factor: 1
+      transaction.state.log.replication.factor: 1
+      transaction.state.log.min.isr: 1
+      log.message.format.version: "2.0.1"
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 3
+    storage:
+      type: ephemeral
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+```
+Apply to cluster:
+
+```bash
+oc apply -f kafka.yaml
+```
+
+After recieving metric objects directly from Kafka, the cruise control metric reporter will filter out undesired metric objects, and store the remaining in a concurrent hash map. Then at periodic intervals designated by the user, the reporter will loop through the concurrent hash map and wrap each metric object's value into a Cruise Control Metric object. The reporter will add information like the metric type, time of when the metric value was recorded, and the broker id of where the metric was taken from. Finally, then Cruise Control metric object is formatted into a byte array and stored back into Kafka within a topic designated for cruise control metics.
```

Comment:
How does this metric reporter work? Is it really so different from the JMX exporter we are using in our Kafka images?
I read here that "...will filter out undesired metric objects..." so it seems to me that this CC metric reporter doesn't get just the metrics that CC needs but many more to not consider.

Reply:
Typo "After receiving..."

Reply:
The Cruise Control metric reporter and other included metric reporters are executed as background threads on every broker at startup. The metric reporters track metric objects that are passed to them by the Kafka nodes and format the stats they receive from these metric objects into specific formats for export to other systems. For example, the Cruise Control metric reporter will transform the received metric objects into a Cruise Control metric objects and store the them in a Kafka topic for later Cruise Control processing. Since Cruise Control doesn't need all of the metric objects passed to it from Kafka for partition balancing, the Cruise Control metric reporter will only format and store the metric objects it needs from Kafka. The metric reporters _prepare_ the metrics for export.  For example, the JMX metric reporter wraps all Kafka metic objects into JMX MBeans so that java agents like the Prometheus JMX exporter(included in the Kafka images) can scrape the JMX MBeans for exporting to other systems like Prometheus.